packages/mock-block-dock/src/use-mock-block-props/use-mock-datastore.ts

```diff
@@ -818,9 +840,110 @@ export const useMockDatastore = (
+
+    const entityTypeInStore = newEntityTypes.find(
+      (entityType) => entityType.entityTypeId === entityTypeId,
+    );
+
+    if (entityTypeInStore) {
+      // If block schema changed, update entity type
+      if (JSON.stringify(blockSchema) !== JSON.stringify(prevBlockSchema)) {
+        Object.assign(entityTypeInStore.schema, blockSchema ?? {});
+        setEntityTypes(newEntityTypes);
+      }
+    } else {
+      // Create a new entity type if the block's entityTypeId isn't in datastore
+      const newBlockEntityType: EntityType = {
+        entityTypeId,
+        schema: {
+          title: "BlockType",
+          type: "object",
+          $schema: "https://json-schema.org/draft/2019-09/schema",
+          $id: `http://localhost/${entityTypeId}`,
+          ...(blockSchema ?? {}),
+        },
+      };
+      setEntityTypes([newBlockEntityType, ...entityTypes]);
+    }
+
+    const newEntities = [...entities];
+    const entityInStore = newEntities.find(
+      (entity) =>
+        entity.entityId === newExternalBlockEntity.entityId &&
+        entity.entityTypeId === newExternalBlockEntity.entityTypeId,
+    );
+
+    if (entityInStore) {
+      // if block entity exists in store and there was no change, do nothing
+      if (
+        JSON.stringify(entityInStore.properties) ===
+        JSON.stringify(newExternalBlockEntity.properties)
+      ) {
+        return;
+      }
+
+      // If the block entity already exists in the store, and
+      // was changed, update what was changed
+      Object.assign(
+        entityInStore.properties,
+        newExternalBlockEntity.properties,
+      );
+      setEntities(newEntities);
+      setBlockEntity(entityInStore);
+    } else {
+      const newBlockEntity = {
+        entityId,
+        entityTypeId,
+        properties: {},
+      };
+
+      if (blockEntity && Object.keys(blockEntity).length > 0) {
+        Object.assign(newBlockEntity.properties, blockEntity.properties);
+      }
+
+      setEntities([newBlockEntity, ...entities]);
+      setBlockEntity(newBlockEntity);
+    }
+  };
+
+  // Update datastore if block isn't in datastore
+  if (
+    !entities.find(
+      (entity) =>
+        entity.entityId === blockEntity.entityId &&
+        entity.entityTypeId === blockEntity.entityTypeId,
+    ) &&
+    [blockEntity.entityId, blockEntity.entityTypeId].every((key) => !!key)
+  ) {
+    updateStore(blockEntity);
+  }
+
+  if (
+    externalBlockEntity !== prevExternalBlockEntity &&
+    JSON.stringify(blockEntity) !== JSON.stringify(externalBlockEntity)
+  ) {
+    const entity = externalBlockEntity ?? ({} as Entity);
+    // This handles scenarios where the passed in block entity doesn't have an
+    // entityId or entityTypeId
+    if (!entity.entityId) {
+      entity.entityId = defaultBlockEntityId;
+    }
+    if (!entity.entityTypeId) {
+      entity.entityTypeId = defaultBlockEntityTypeId;
+    }
+    setBlockEntity(entity);
+  }
```

Comment:
Block Entity can be updated either
1. externally (e.g via block properties tab in BP hub) or
2. internally (e.g when user switches block entity)

This takes account of the different sources of changes and adds a default entity(Id/typeId) if it's not present